flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/AsyncDeleteFiles.java

```diff
@@ -0,0 +1,106 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.maintenance.operator;
+
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.concurrent.ExecutorService;
+import java.util.function.Predicate;
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.metrics.Counter;
+import org.apache.flink.streaming.api.functions.async.ResultFuture;
+import org.apache.flink.streaming.api.functions.async.RichAsyncFunction;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.flink.TableLoader;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.util.ThreadPools;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Delete the files using the {@link FileIO}. */
+@Internal
+public class AsyncDeleteFiles extends RichAsyncFunction<String, Boolean> {
+  private static final Logger LOG = LoggerFactory.getLogger(AsyncDeleteFiles.class);
+  public static final Predicate<Collection<Boolean>> FAILED_PREDICATE = new FailedPredicate();
+
+  private final String name;
+  private final FileIO io;
+  private final int workerPoolSize;
+  private final String tableName;
+
+  private transient ExecutorService workerPool;
+  private transient Counter failedCounter;
+  private transient Counter succeededCounter;
+
+  public AsyncDeleteFiles(String name, TableLoader tableLoader, int workerPoolSize) {
+    Preconditions.checkNotNull(name, "Name should no be null");
+    Preconditions.checkNotNull(tableLoader, "Table loader should no be null");
+
+    this.name = name;
+    tableLoader.open();
+    Table table = tableLoader.loadTable();
+    this.io = table.io();
+    this.workerPoolSize = workerPoolSize;
+    this.tableName = table.name();
+  }
+
+  @Override
+  public void open(Configuration parameters) throws Exception {
+    this.failedCounter =
+        getRuntimeContext()
+            .getMetricGroup()
+            .addGroup(TableMaintenanceMetrics.GROUP_KEY, name)
+            .counter(TableMaintenanceMetrics.DELETE_FILE_FAILED_COUNTER);
+    this.succeededCounter =
+        getRuntimeContext()
+            .getMetricGroup()
+            .addGroup(TableMaintenanceMetrics.GROUP_KEY, name)
+            .counter(TableMaintenanceMetrics.DELETE_FILE_SUCCEEDED_COUNTER);
+
+    this.workerPool =
+        ThreadPools.newWorkerPool(tableName + "-" + name + "-async-delete-files", workerPoolSize);
+  }
+
+  @Override
+  public void asyncInvoke(String fileName, ResultFuture<Boolean> resultFuture) {
+    workerPool.execute(
+        () -> {
+          try {
+            LOG.info("Deleting file: {} with {}", fileName, name);
+            io.deleteFile(fileName);
```

Comment:
we should use the batch delete API from `SupportsBulkOperations`

Reply:
If we want to use the batch delete feature then we need state handling too to store the not yet delete files.
Also theoretically not all FileIO implements the `SupportsBulkOperations` interface.

Reply:
Most `FileIO` implementations support the bulk deletes. We should definitely leverage the bulks deletion when `FileIO` supports it. It makes a huge difference.

<img width="387" alt="image" src="https://github.com/user-attachments/assets/f192454c-40b5-4dd5-a556-a3cac92adc80">

Reply:
I am wondering if we still need async if leveraging bulk deletions, which are available for all major implementations Azure, GCP, Hadoop, S3.

I don't know we need to store state. during the checkpoint, the operator can flush all pending files and submit a synchronous bulk deletion call. This is like the file flushing for `IcebergStreamWriter#prepareSnapshotPreBarrier`.

Reply:
This is the javadoc for the prepareSnapshotPreBarrier:
```
Important: This method should not be used for any actual state snapshot logic, because it will inherently be within the synchronous part of the operator's checkpoint. If heavy work is done within this method, it will affect latency and downstream checkpoint alignments.
```

Deleting high number of files could be quite heavy even for bulk deletes.

So I switched to the bulk delete implementation and added a `deleteBatchSize` to limit the time spent on deletes.
The delete parallelism is also really dependent on the use-case, so added another config `deleteParallelism`

Reply:
there is really no fundamental difference whether we use `prepareSnapshotPreBarrier` or not. if we use the bulk delete API, the check barrier can come right after a bulk delete request with or without `prepareSnapshotPreBarrier`. The main thing is that batch delete requests have higher latency. 

Async model can increase the throughput with single file deletion. but it is overall inefficient and more expensive. It also can't avoid the potential throttling problem.

The change seems not pushed up to the PR yet. I imagine `deleteBatchSize` is used for buffering (like the target file size for rolling file writer). It can be set to 1,000 which is the S3 limit. Wait until 1,000 files or checkpoint barrier.

Reply:
Pushed the change.